funnel/views/helpers.py

```diff
@@ -20,3 +24,64 @@ def localize_date(date, from_tz=utc, to_tz=utc):
             to_tz = pytz_timezone(to_tz)
         return from_tz.localize(date).astimezone(to_tz).replace(tzinfo=None)
     return date
+
+def basepath(url):
+    """
+    Returns the base path of a given a URL
+    Eg::
+
+        basepath("https://hasgeek.com/1")
+        >> u"https://hasgeek.com
+
+    :param url: A valid URL unicode string. Eg: https://hasgeek.com
+    """
+    parsed_url = urlparse(url)
+    if not (parsed_url.scheme or parsed_url.netloc):
+        raise ValueError("Invalid URL")
+    return u"{scheme}://{netloc}".format(scheme=parsed_url.scheme, netloc=parsed_url.netloc)
+
+def cors(any_get=False):
+    """
+    Adds CORS headers to the decorated view function.
+
+    Requires `app.config['ALLOWED_ORIGINS']` to be defined with a list
+    of permitted domains. Eg: app.config['ALLOWED_ORIGINS'] = ['https://example.com']
+
+    :param any_get: Specify * on the Access-Control-Allow-Origin header for any GET request
+    """
+    def inner(f):
+        def add_headers(resp, origin):
+            resp.headers['Access-Control-Allow-Origin'] = origin
+            resp.headers['Access-Control-Allow-Methods'] = 'POST, OPTIONS, GET'
+            # echo the request's headers
+            resp.headers['Access-Control-Allow-Headers'] = request.headers.get('Access-Control-Request-Headers')
+            # debugging only
+            if app.debug:
+                resp.headers['Access-Control-Max-Age'] = '1'
+            return resp
+
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            origin = request.headers.get('Origin')
+            if not origin:
+                # Firefox doesn't send the Origin header, so read the Referer header instead
+                # TODO: Remove this conditional when Firefox starts adding an Origin header
+                referer = request.referrer
+                if referer:
+                    origin = basepath(referer)
```

Comment:
Is Firefox's behavioral documented, and is this the recommended workaround? It's a potential security hole otherwise.